Tutorial 6 Convert

F20170604 opened the pull request:
Complete Tutorial with Image Augmentation added.

Original Tutorial Link: https://github.com/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l05c02_dogs_vs_cats_with_augmentation.ipynb

S4TF Tutorial Colab File Link: https://colab.research.google.com/drive/148B0g-zaefmDTWxbEk1OiKGCEPojWwgz

Original Tutorial Graph:
![Original](https://user-images.githubusercontent.com/31190928/61287622-e9e42000-a7e2-11e9-9bba-79930afddf0e.png)

S4TF Tutorial Graph:
![S4TF-Tut-6](https://user-images.githubusercontent.com/31190928/61287649-f5374b80-a7e2-11e9-8a57-e65a38f2be3a.png)

F20170604 committed:
Tutorial 6 Convert

Complete Tutorial with Image Augmentation added.

Original Tutorial Link: https://github.com/tensorflow/examples/blob/master/courses/udacity_intro_to_tensorflow_for_deep_learning/l05c02_dogs_vs_cats_with_augmentation.ipynb

S4TF Tutorial Colab File Link: https://colab.research.google.com/drive/148B0g-zaefmDTWxbEk1OiKGCEPojWwgz

Ayush517 commented:
I missed on adding the documentation and steps in the tutorial. I'll add them right away.

F20170604 committed:
Added documentation

F20170604 committed:
Update S4TF_Tutorial_6.ipynb

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "        var convolved3 = pool3(conv1c(convolved2))\n",
+        "        var convolved4 = pool4(conv1d(convolved3))\n",
+        "        return layer1b(layer1a(flatten(dropout1a(convolved4))))\n",
+        "        \n",
```
Remove redundant line.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      },
+      "source": [
+        "// Function `tensor(path:ImageAumentation:)` returns the resized image tensor w/ augmentation and it's corresponding label.\n",
+        "func tensor(path: String, ImageAumentation: Bool) -> (Tensor<Float>, Int32){\n",
```
* All argument labels need to use camel case.
* Typo: "augmentation", not "aumentation".
* Please rename this to `augmented: Bool`.
* Please replace all occurrences of `ImageAumentation:` with `augmented:`.

rxwei replied:
1. Add a space between `(` and `{`.
2. The function base name "tensor" suggests it returns a single `Tensor`, but it also returns an `Int32`. And the base name does not say anything informative about what this function does. Please rename the function to `resizedImage(fromPath:augmented:)`, and add tuple element labels to the return type.

```suggestion
        "func resizedImage(fromPath path: String, augmented: Bool) -> (image: Tensor<Float>, label: Int32) {\n",
```

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "    var image = np.array(img, dtype: np.float32) * (1.0 / 255)\n",
+        "    var imageTensor = Tensor<Float>(numpy: image)!\n",
+        "    \n",
+        "    if ImageAumentation == true {\n",
```
Please use boolean values directly instead of comparing them with boolean values.

```suggestion
        "    if augmented {\n",
```

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "// Function `tensors(fromList:valueCount:ImageAumentation:)` returns the entire list converted to tensors of images and labels from `tensor(path:ImageAumentation:)` function.\n",
+        "func tensors(fromList: PythonObject, valueCount: Int, ImageAumentation: Bool) -> (Tensor<Float>, Tensor<Int32>) {\n",
+        "    let batchFiles = fromList[0 ..< valueCount]\n",
+        "    var labels = [Int32]()\n",
```
Prefer literal initialization for arrays.
```suggestion
        "    var labels: [Int32] = []\n",
```

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "colab": {}
+      },
+      "source": [
+        "func imageAugmentations(path: String) -> Tensor<Float> {\n",
```
```suggestion
        "func argumentedImage(fromPath path: String) -> Tensor<Float> {\n",
```

F20170604 replied:
@rxwei Is the name really supposed to `argumentedImage` or is this a typo and instead you meant `augmentedImage`?

rxwei replied:
Sorry, it's a typo. `augmentedImage`.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "Images must be formatted into appropriately pre-processed floating point tensors before being fed into the network. The steps involved in preparing these images are:\n",
+        "\n",
+        "1. Read images from the disk\n",
```
Please end each sentence with a period.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "    \n",
+        "    image = np.array(image, dtype: np.float32)\n",
+        "    \n",
+        "    let imageTensor = Tensor<Float>(numpy: image)!\n",
```
Return this directly instead of creating a new `let`.

F20170604 replied:
Returning Numpy Array directly was causing OOM or something like that. Further, it was misbehaving and returning out all black images. That's why I am returning image tensor which is easier to handle, as well as can be managed easily in the `resizedImage` function.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "Overfitting often occurs when we have a small number of training examples. One way to fix this problem is to augment our dataset so that it has sufficient number and variety of training examples. Data augmentation takes the approach of generating more training data from existing training samples, by augmenting the samples through random transformations that yield believable-looking image ...
+        "\n",
+        "In S4TF we can implement this using the different Python Libraries through Python Interoperability. We can simply pass different transformations we would want to our dataset as a form of arguments and it will take care of applying it to the dataset during our training process."
```
Always use "Swift for TensorFlow" instead of "S4TF" in any tutorials or official documentation.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+      "source": [
+        "The `tensor` function takes 2 inputs are:\n",
+        "1. Image path as input\n",
+        "2. Whether or not to apply Image Augmentations\n",
```
It's not a proper noun. Use lower case.
```suggestion
        "2. Whether or not to apply image augmentations\n",
```

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "colab_type": "text"
+      },
+      "source": [
+        "The `tensor` function takes 2 inputs are:\n",
```
Always refer to functions by their full name.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "1. Image path as input\n",
+        "2. Whether or not to apply Image Augmentations\n",
+        "\n",
+        "The `tensors` function takes 3 inputs :\n",
```
Use function full name.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "The `tensors` function takes 3 inputs :\n",
+        "\n",
+        "1. List of paths\n",
```
End each item with a period.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "\n",
+        "    // Load the rest images.\n",
+        "    var numberOfFilesDone = 1\n",
+        "    for file in batchFiles[1 ..< valueCount] {\n",
```
```suggestion
        "    for file in batchFiles[1..<valueCount] {\n",
```

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "}\n",
+        "\n",
+        "// Function `tensors(fromList:valueCount:ImageAumentation:)` returns the entire list converted to tensors of images and labels from `tensor(path:ImageAumentation:)` function.\n",
+        "func tensors(fromList: PythonObject, valueCount: Int, ImageAumentation: Bool) -> (Tensor<Float>, Tensor<Int32>) {\n",
```
```suggestion
        "func images(fromList: PythonObject, valueCount: Int, augmented: Bool) -> (images: Tensor<Float>, labels: Tensor<Int32>) {\n",
```

Also, what does the "value" in `valueCount` stand for? Could you use a more descriptive name?

F20170604 replied:
Is `tensorCount` more appropriate? Or `imageCount`?

rxwei replied:
`imageCount` is more descriptive here.

rxwei commented on S4TF_Tutorial_6.ipynb:
```diff
@@ -0,0 +1,1356 @@
+        "}\n",
+        "\n",
+        "// The training loop.\n",
+        "for epoch in 1 ... epochCount {\n",
```
```suggestion
        "for epoch in 1...epochCount {\n",
```

F20170604 committed:
Changes made proposed by Richard

Redundant Line removed
Labels renamed
Using boolean values directly
Function names corrected

F20170604 committed:
Minor Changes

Forgot to rename the `tensors(fromList:imageCount:augmented:)` function to `images(fromList:imageCount:augmented:)`.
Renamed it in this commit.

Ayush517 commented:
@rxwei I've made the changes in this tutorial as well as in the next 2 tutorials too.

Ayush517 commented:
Hey @rxwei, Any more changes in this notebook?
